README.md
# Chainlink external adapter for Quandl API

![Test, Build and push Docker image](https://github.com/willianpaixao/cl-ea-quandl/workflows/Test,%20Build%20and%20push%20Docker%20image/badge.svg?branch=master)
![CodeQL](https://github.com/willianpaixao/cl-ea-quandl/workflows/CodeQL/badge.svg?branch=master)
![Snyk](https://github.com/willianpaixao/cl-ea-quandl/workflows/Snyk%20vulnerability%20scan/badge.svg?branch=master)

## Getting started
**REQUIRED:** First step is creating an account at [Quandl](https://www.quandl.com/sign-up) and creating an API key.

Once you have it, simply run the image:
``` bash
$ docker container run --rm --detach --name cl-ea-quandl --publish 5000:5000 --env QUANDL_API_KEY=[INSERT HERE YOUR KEY] willianpaixao/cl-ea-quandl:latest
```
> NOTE: don't forget to replace with your API key

To quickly test, make a request using `curl`:
``` bash
$ curl --request POST --header "Content-Type: application/json" --data '{"id": 1, "data": {"dataset": "FRED/GDP"}}' localhost:5000
{
  "data": {
    "dataset": "FRED/GDP",
    "result": 21157.122
  },
  "jobRunID": 1,
  "statusCode": 200
}
```

## Quandl API
Now that you have a running adapter, look at the [Quandl API documentation](https://docs.quandl.com) for all possible parameters that can be passed
and at the [Quandl feed explorer](https://www.quandl.com/search) to pick the data you need.

### Examples
You can explore the free (as in "free of charge") data feeds [here](https://www.quandl.com/search?filters=%5B%22Free%22%5D).
Most common usages are for macroeconomic data and financial indexes, some examples are:

| Dataset           | Description                      |
| ----------------- | -------------------------------- |
|`FRED/GDP`         | USA's Gross Domestic Product     |
|`FRED/UNRATE`      | USA's Civilian Unemployment Rate |
|`FRED/GFDEBTN`     | USA's Federal Public Debt        |
|`NASDAQOMX/OMXS30` | OMX Stockholm 30 Index           |
|`BCB/11`           | Brazilian interest rate          |

> NOTE: You can purchase a premium feed but this is out of the scope of this service.

## Development
You can easily build with Docker, by running:
``` bash
$ docker image build --tag cl-ea-quandl:latest .
```

``` bash
$ gcloud functions deploy cl-ea-quandl --set-env-vars QUANDL_API_KEY=[INSERT HERE YOUR KEY] --entry-point index --runtime python38 --memory 128MB --trigger-http --allow-unauthenticated
```

Alternatively you can run locally without Docker with:
``` bash
$ QUANDL_API_KEY=[INSERT HERE YOUR KEY] FLASK_ENV=development flask run
```

### Testing
Make sure you have installed all dependencies using `Pipfile`, then run:
```
$ FLASK_ENV=testing pytest --setup-show
```

### Contributing
Did you run into a bug? Please create an issue.
Do you want to contribute? All PRs are welcome.


CHANGELOG.md
# Changelog

## [v1.0](https://github.com/willianpaixao/cl-ea-quandl/releases/tag/v1.0) - 2020-12-06
First stable version. Basic features are:
* Docker image published to Docker hub at [willianpaixao/cl-ea-quandl](https://hub.docker.com/repository/docker/willianpaixao/cl-ea-quandl)
* Github Actions used for building, testing and deployment
* Environmental variables used to fetch `QUANDL_API_KEY` on runtime
* `production`, `development` or `testing` can be set for `FLASK_ENV` to setup your application environment
* `dataset` and `rows` are the only parameter accepted at the moment
* Pytest is used to run a few raw tests
* Among other features.This is a preliminary release that should receive more documentation and tests soon


tests/conftest.py
import pytest

from app import create_app


@pytest.fixture
def app():
    """Create and configure a new app instance for each test."""
    app = create_app()
    yield app


@pytest.fixture
def client(app):
    """A test client for the app."""
    return app.test_client()


tests/test_app.py
import pytest

from app import create_app


def test_config(app):
    """Test create_app without passing test config."""
    assert create_app().testing


def test_call_adapter(client):
    rv = client.post('/', json={'id': '1', 'data': {}})
    json_data = rv.get_json()
    assert json_data.get('statusCode') == 500
    assert json_data.get('status') == 'errored'


def test_request_fred(client):
    rv = client.post('/', json={"id": 1, "data": {"dataset": "FRED/GDP"}})
    json_data = rv.get_json()
    assert json_data.get('statusCode') == 200
    assert 'result' in json_data['data']
    rv = client.post('/', json={"id": 2, "data": {"dataset": "FRED/UNRATE", "rows": 2}})
    json_data = rv.get_json()
    assert json_data.get('statusCode') == 200
    assert 'result' in json_data['data']


def test_extra_params(client):
    rv = client.post('/', json={"id": "01acef78e00f45ca911f9202124fbabf",
                                "data": {
                                    "address": "0x795FB736ef447f649EBD462F35d6cC8437925fC0",
                                    "dataset": "FRED/UNRATE"
                                }})
    json_data = rv.get_json()
    assert json_data.get('statusCode') == 200
    assert 'result' in json_data['data']


if __name__ == '__main__':
    # Calling pytest directly instead of the CLI
    pytest.main()


app.py
import logging
import os

from flask import Flask, request


def create_app(test_config=None):
    """Create and configure an instance of the Flask application."""
    app = Flask(__name__, instance_relative_config=True)
    if test_config is None:
        configure_app(app)
    else:
        app.config.update(test_config)

    @app.before_request
    def log_request_info():
        app.logger.debug('Headers: %s', request.headers)
        app.logger.debug('Body: %s', request.get_data())

    from main import api
    app.register_blueprint(api)

    @app.route('/healthcheck', methods=['GET'])
    def healthcheck():
        return 'OK'

    return app


def configure_app(app):
    """  Loads a different set of configurations depending on the desired environment

    :param app: Flask app instance
    """
    app.config.from_object('config.ProductionConfig')
    if app.config['ENV'] == 'development':
        app.config.from_object('config.DevelopmentConfig')
    elif app.config['ENV'] == 'testing':
        app.config.from_object('config.TestingConfig')

    if 'SECRETS_FILE' in os.environ:
        try:
            app.config.from_envvar('SECRETS_FILE')
        except SyntaxError:
            app.logger.error("Invalid secret file '{}', aborting...", os.environ['SECRETS_FILE'])
    elif 'QUANDL_API_KEY' in os.environ:
        try:
            app.config['QUANDL_API_KEY'] = os.environ['QUANDL_API_KEY']
        except SyntaxError:
            app.logger.error("Invalid Quandl API key '{}', aborting...", os.environ['QUANDL_API_KEY'])
    else:
        app.logger.error("Required Quandl API key not provided, aborting...")

    app.logger.info("Configuration loaded successfully")

    # The following libraries gives excessive logging, so fine tuning is needed
    logging.getLogger("urllib3").setLevel(logging.WARNING)


if __name__ == '__main__':
    app = create_app()
    app.run(host="0.0.0.0")


bridge.py
import os

import quandl

from flask import current_app


class Bridge:

    def __init__(self):
        if 'QUANDL_API_KEY' in current_app.config:
            quandl.ApiConfig.api_key = current_app.config['QUANDL_API_KEY']
        else:
            quandl.ApiConfig.api_key = os.environ['QUANDL_API_KEY']

    def request(self, data):
        """

        :param data: JSON Object to feed Quandl API request
        :return: JSON object with the response from he API
        """
        try:
            rows = 1
            if 'rows' in data:
                rows = data.get('rows')
            df = quandl.get(dataset=data.get('dataset'), rows=rows)
            if df['Value'].size == 1:
                return df['Value'].get(0)
            else:
                return df['Value'].to_list()
        except Exception as e:
            raise e


config.py
from logging.config import dictConfig


class Config(object):
    DATABASE_URI = ':memory:'
    DEBUG = False
    SESSION_COOKIE_SECURE = True
    TESTING = False

    dictConfig({
        'version': 1,
        'disable_existing_loggers': True,
        'formatters': {
            'default': {
                'format': '[%(asctime)s] %(levelname)s: %(message)s',
            }
        },
        'handlers': {
            'wsgi': {
                'class': 'logging.StreamHandler',
                'stream': 'ext://flask.logging.wsgi_errors_stream',
                'formatter': 'default'
            }
        },
        'root': {
            'level': 'INFO',
            'handlers': ['wsgi']
        }
    })


class DevelopmentConfig(Config):
    DEBUG = True
    SESSION_COOKIE_SECURE = False

    dictConfig({
        'version': 1,
        'disable_existing_loggers': True,
        'formatters': {
            'default': {
                'format': '[%(asctime)s] %(levelname)s: %(message)s',
            }
        },
        'handlers': {
            'wsgi': {
                'class': 'logging.StreamHandler',
                'stream': 'ext://flask.logging.wsgi_errors_stream',
                'formatter': 'default'
            }
        },
        'root': {
            'level': 'DEBUG',
            'handlers': ['wsgi']
        }
    })


class ProductionConfig(Config):
    pass


class TestingConfig(Config):
    SESSION_COOKIE_SECURE = False
    TESTING = True


adapter.py
from schema import Schema, And, Optional
from bridge import Bridge


class Adapter:

    def __init__(self, input):
        self.id = input.get('id', '1')
        self.request_data = input.get('data')
        if self.validate_request_data():
            self.bridge = Bridge()
            self.create_request()
        else:
            self.result_error('Invalid data provided')

    def validate_request_data(self):
        """ Validates the incoming data against a pre-defined schema

        :return: True if the data provided is a valid JSON containing a
        dataset key, False otherwise
        """
        schema = Schema({'dataset': str,
                         Optional('rows'): And(int, lambda n: 0 < n)},
                        ignore_extra_keys=True)
        return schema.is_valid(self.request_data)

    def create_request(self):
        try:
            data = self.request_data
            data['result'] = self.bridge.request(self.request_data)
            self.result_success(data)
        except Exception as e:
            self.result_error(e)

    def result_success(self, data):
        self.result = {
            'jobRunID': self.id,
            'data': data,
            'statusCode': 200,
        }

    def result_error(self, error):
        self.result = {
            'jobRunID': self.id,
            'status': 'errored',
            'error': f'There was an error: {error}',
            'statusCode': 500,
        }


main.py
from flask import Blueprint, jsonify, request
from schema import Schema, And, Optional

from bridge import Bridge

api = Blueprint('api', __name__)


def index(request):
    data = request.get_json()
    if data == '':
        data = {}
    adapter = Adapter(data)
    return jsonify(adapter.result), adapter.result['statusCode']


@api.route('/', methods=['POST'])
def call_adapter():
    data = request.get_json()
    if data == '':
        data = {}
    adapter = Adapter(data)
    return jsonify(adapter.result), adapter.result['statusCode']


class Adapter:

    def __init__(self, input):
        self.id = input.get('id', '1')
        self.request_data = input.get('data')
        if self.validate_request_data():
            self.bridge = Bridge()
            self.create_request()
        else:
            self.result_error('Invalid data provided')

    def validate_request_data(self):
        """ Validates the incoming data against a pre-defined schema

        :return: True if the data provided is a valid JSON containing a
        dataset key, False otherwise
        """
        schema = Schema({'dataset': str,
                         Optional('rows'): And(int, lambda n: 0 < n)},
                        ignore_extra_keys=True)
        return schema.is_valid(self.request_data)

    def create_request(self):
        try:
            data = self.request_data
            data['result'] = self.bridge.request(self.request_data)
            self.result_success(data)
        except Exception as e:
            self.result_error(e)

    def result_success(self, data):
        self.result = {
            'jobRunID': self.id,
            'data': data,
            'statusCode': 200,
        }

    def result_error(self, error):
        self.result = {
            'jobRunID': self.id,
            'status': 'errored',
            'error': f'There was an error: {error}',
            'statusCode': 500,
        }
